PR - Integrate approved API SUPs

ajcraig opened the pull request:
### Description

This pull request completes the integration of two approved SUPs into the specification. The SUPs that were integrated:
- https://github.com/margo/specification-enhancements/blob/main/submitted/sup_desired_state_via_manifest_api.md
- https://github.com/margo/specification-enhancements/blob/main/submitted/api-details-finalization-SUP-folder/api-details-finalization-SUP-WIP.md

#### Issues Addressed

List and link all the issues addressed by this PR.
- https://github.com/margo/specification/issues/99
- https://github.com/margo/specification/issues/100
- https://github.com/margo/specification/issues/108
- https://github.com/margo/specification/issues/97
- https://github.com/margo/specification/issues/101

#### Change Type

Please select the relevant options:

- [ ] Fix (change that resolves an issue)
- [x] New enhancement (change that adds specification content)
- [ ] Content edits (change that edits existing content)

#### Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) document.
- [x] My changes adhere to the established patterns, and best practices.

Commit:
small deployment status updates
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Initial content injection from approved SUP
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

matlec committed:
feat: Migrate Desired State SUP into spec

Signed-off-by: Matthias Lechner <matlec.public@gmail.com>

matlec committed:
feat: Add note on digest format vs RFC 9421

Signed-off-by: Matthias Lechner <matlec.public@gmail.com>

Commit:
Updated deployment status and device capabilities markdowns.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

matlec committed:
feat: Update workload deployment concepts

Signed-off-by: Matthias Lechner <matlec.public@gmail.com>

Commit:
Improvements to concepts, certificate and onboarding API.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Small commit to reorder the navigation to match the client/server interaction workflow.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Updates to the various markdowns associated with the onboarding, capabilities, and certificate API sections. Additional improvements to the concepts section.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Addressed feedback received via internal review.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Embedded the management interface swagger definition.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Pinned python 3.12 in the pages job. There was a compatibility issue with 3.x being used with Py03.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

matlec commented on system-design/specification/margo-management-interface/device-capabilities.md:
```diff
@@ -1,28 +1,40 @@
 # Device Capabilities
 
-Devices MUST provide the Workload Fleet Management service with its capabilities and characteristics. This is done by calling the Device API's `device capabilities` endpoint.
+Devices MUST provide the Workload Fleet Management service with its capabilities and characteristics. This is done by calling the Device API's `device capabilities` endpoint. Reporting the device capabilities is the final step in the onboarding of the device's client. 
```
> Reporting the device capabilities is the final step in the onboarding of the device's client.
While this is technically correct, devices may report updated capabilities upon change.

ajcraig replied:
I think the note on line 5 covers the additional updates required post onboarding.

matlec commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+#### API Port Details
+
+This API is designed to minimize the ports required on the customer's infrastructure to enable cloud to edge communication. 
+    -The API MUST ONLY utilize port 443 for it's traffic. 
```
```suggestion
    -The API MUST ONLY utilize port 443 for its traffic. 
```

matlec commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+- This client ID MAY be in the format of UUIDv4
+- Other URL safe string identifiers are permitted to be used
+
+> Note: This does introduce a change in philosophy regarding the unique identifier, changing towards a more flexible client id that is separate from device id. 
```
Should we leave this note in here? Not sure if people that are not aware of earlier revisions of the draft spec would understand this note.

ajcraig replied:
Removed the note in latest commit.

phil-abb replied:
yeah, I don't think this is really needed either.

matlec commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+> Note: This does introduce a change in philosophy regarding the unique identifier, changing towards a more flexible client id that is separate from device id. 
+
+#### Certificate Information
+Both the WFM and Device client MUST utilize a x.509 certificates to represent themselves within the API interactions.
```
```suggestion
Both the WFM and Device client MUST utilize X.509 certificates to represent themselves within the API interactions.
```

matlec commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
-> Action: Ideally this URL is discoverable instead of having to manually enter it but we still need to determine if there is a good way to make this discoverable by using something like the FDO Rendezvous service or multicast DNS. Also, once we determine how the Margo compliant device onboarding and fleet management is going to work it will probably impact this.
-
-To ensure the management client is configured to communicate with the correct Workload Fleet Management web service, the device's management client needs to be configured with the expected URL. The device vendor MUST provide a way for the end user to manually set the URL the device's management client uses to communicate with the Workload FLeet Management solution chosen by the end user. 
+- The device's management client uses the [Onboarding API](../../specification/margo-management-interface/certificate-api.md) to onboard with the Workload Fleet Management service by providing it's x.509 certificate
```
```suggestion
- The device's management client uses the [Onboarding API](../../specification/margo-management-interface/certificate-api.md) to onboard with the Workload Fleet Management service by providing its X.509 certificate
```

matlec commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
-
-To ensure the management client is configured to communicate with the correct Workload Fleet Management web service, the device's management client needs to be configured with the expected URL. The device vendor MUST provide a way for the end user to manually set the URL the device's management client uses to communicate with the Workload FLeet Management solution chosen by the end user. 
+- The device's management client uses the [Onboarding API](../../specification/margo-management-interface/certificate-api.md) to onboard with the Workload Fleet Management service by providing it's x.509 certificate
+- The device's management client receives it's unique client Id assigned via the Workload Fleet Manager
```
```suggestion
- The device's management client receives its unique client Id assigned via the Workload Fleet Manager
```

matlec commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
+    Server-->>Client: 201 Created
+    Note over Client,Server: Enrollment is complete
+
+    Note over Client,Server: Assumed Desired state has been transmitted via WFM(out of scope for this SUP)
```
We should avoid mentioning SUP in the final spec

ajcraig replied:
Removed in latest commit.

matlec commented on system-design/specification/margo-management-interface/device-onboarding.md:
````diff
@@ -1,47 +1,102 @@
-POST /onboarding/
+```json
+{
+    "clientId": "<base-64 encoded UUID>"
````
In the API security section, use of UUID is not enforced, this needs to be clarified in the spec

ajcraig replied:
I attempted to correct/match the security section with this example.

matlec reviewed (changes requested):
General:
1. Is there any rationale for how we choose to which `apiVersion` a specific Margo resource belongs? Today, it looks like each kind gets its own apiVersion which defeats the purpose of having `apiVersion` properties in the first place.

Commit:
Update system-design/specification/margo-management-interface/api-requirements-and-security.md

Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Co-authored-by: Matthias Lechner <matlec.public@gmail.com>
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Update system-design/specification/margo-management-interface/api-requirements-and-security.md

Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Co-authored-by: Matthias Lechner <matlec.public@gmail.com>
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Update system-design/specification/margo-management-interface/device-onboarding.md

Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Co-authored-by: Matthias Lechner <matlec.public@gmail.com>
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Update system-design/specification/margo-management-interface/device-onboarding.md

Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Co-authored-by: Matthias Lechner <matlec.public@gmail.com>
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

phil-abb commented on system-design/concepts/workload-fleet-managers/device-client-onboarding.md:
The Page name should probably be "device-client-onboarding.md"

ajcraig replied:
resolved

phil-abb commented on system-design/specification/margo-management-interface/management-interface-swagger.md:
The page name should probably be "management-interface-swagger.md"

ajcraig replied:
resolved

phil-abb commented on src/specification/margo-management-interface/desired-state.linkml.yaml:
```diff
@@ -2,9 +2,8 @@
-  [Kubernetes custom resource definition](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
-  and made available to the device's management client as a YAML document using the OpenGitOps pattern.
+  Each workload is represented as an `ApplicationDeployment` YAML file that specifies its components, configuration, and parameters.
+  This definition is shared across the Desired State and Deployment Status APIs.
```
I'm not sure I understand this statement. There is no common definition shared between the Desired State and the Deployment  Status

matlec replied:
You are right, this statement is not accurate. I tried to improve the phrasing to describe the link between the two APIs.

Commit:
Updates based on recent review
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

phil-abb commented on src/specification/margo-management-interface/resources/index.md.jinja2:
The styles on this page are a lot different than the other pages. Not sure if we want to do anything about it now or not, since we haven't defined a style guide yet.

matlec replied:
Agreed. I tried to make the documentation style of the desire state API to be closer to the style we use throughout the rest of the spec.

phil-abb commented on src/specification/margo-management-interface/resources/index.md.jinja2:
It feels like we're merging two concepts onto a single page. While the concepts are related, they serve different purposes:
1. Definition for the "ApplicationDeployment"
2. Obtaining the desired state

I think these should be two different pages. At a minimum, I think we need to flip the order so we explain what the "ApplicationDeployment" is first, and then explain how it is retrieved.

matlec replied:
In my opinion, the desired state API and the `ApplicationDeployment` schema are tightly coupled components of a single synchronization protocol, not independent concepts. The document follows a typical client implementation workflow: retrieve, verify, then parse. Based on that, I'd rather keep those descriptions on a single page.

Are there other opinions on this topic?

phil-abb replied:
I'm ok with keeping them on the same page, but I still think it should be flipped so people know what an ApplicationDeployment is - probably not too important for PR1, so I won't press the issue :D

ajcraig replied:
We agreed to keep this the same for PR1 and address after.

phil-abb commented on src/specification/margo-management-interface/resources/index.md.jinja2:
```diff
@@ -1,12 +1,272 @@
+
+| Code | Description |
+| ----- | ---------- |
+| **200 OK** | The response body contains the manifest. The server **MUST** include a valid `ETag` and `Content-Type: application/vnd.margo.manifest.v1+json`. |
```
What is the expected response if this call is made before any workloads/applications are configured to be deployed? 

This case kind of gets addressed below, but I think it would be good to state it here.

matlec replied:
Regardless of the number of workloads to be deployed, the response will always be 200 in case of success. Having zero workloads is no special case. The response fields below describe the exact shape of the body in the case of zero workloads.

phil-abb replied:
It's fine for now.

phil-abb commented on src/specification/margo-management-interface/resources/index.md.jinja2:
```diff
@@ -1,12 +1,272 @@
+
+  - If the manifest is unchanged, the WFM responds with **304 Not Modified**.
+  - If the manifest has changed (**200 OK**), the client:
+    - Verifies the `manifestVersion` is strictly greater than the stored version. If not, the update **MUST** be rejected and logged as a security event.
```
Should we put a note here to indicate that what "logged as a security event" entails hasn't been defined yet?

matlec replied:
Thanks, good input. I relaxed the requirement on logging and added a note, please check

phil-abb commented on system-design/concepts/workload-fleet-managers/client-onboarding.md:
```diff
@@ -0,0 +1,41 @@
+# Device Client Onboarding
+
+To enable workload management, the device's client must first establish trust and complete an onboarding process with the End Users selected Workload Fleet Manager. This onboarding process enables late binding, which is a critical Margo non-functional requirement that enables a device to bind to any Margo compatible Workload Fleet Manager. 
```
```suggestion
To enable workload management, the device's client first establishes trust and completes an onboarding process with the End Users' selected Workload Fleet Manager. This onboarding process enables late binding, which is a critical Margo non-functional requirement that enables a device to bind to any Margo-compatible Workload Fleet Manager. 
```

phil-abb replied:
I think we should try to avoid using MUST, MAY, SHALL, SHOULD, etc. terms in the pages that are not specification pages.

phil-abb commented on system-design/concepts/workload-fleet-managers/client-onboarding.md:
```diff
@@ -0,0 +1,41 @@
+## Trust Establishment
+
+Initial trust is established between the device's Workload Fleet Management (WFM) Client and the WFM using server-side TLS.
+Before the WFM Client can connect securely, it must obtain the WFM's root CA certificate. This trust anchor may be:
```
```suggestion
Before the WFM Client can connect securely, it obtains the WFM's root CA certificate. This trust anchor can be:
```

phil-abb commented on system-design/concepts/workload-fleet-managers/client-onboarding.md:
```diff
@@ -0,0 +1,41 @@
+Before the WFM Client can connect securely, it must obtain the WFM's root CA certificate. This trust anchor may be:
+
+- downloaded via the Certificate API, provided that an existing trusted channel is available, or
+- delivered out-of-band (e.g. preloaded by the device owner or transferred via USB)
```
```suggestion
- delivered out-of-band (e.g., preloaded by the device owner or transferred via USB)
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+#### API Port Details
+
+This API is designed to minimize the ports required on the customer's infrastructure to enable cloud to edge communication. 
+    -The API MUST ONLY utilize port 443 for its traffic. 
```
```suggestion
    - The API MUST ONLY utilize port 443 for its traffic. 
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+
+#### Unique Identifiers
+
+The WFM MUST create create a URL safe client id to uniquely identify each client within the architecture. 
```
```suggestion
The WFM MUST create a URL-safe client ID to uniquely identify each client within the architecture. 
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+- This client ID MAY be in the format of UUIDv4
+- Other URL safe string identifiers are permitted to be used
+
+> Note: This does introduce a change in philosophy regarding the unique identifier, changing towards a more flexible client id that is separate from device id. 
```
```suggestion
> Note: This does introduce a change in philosophy regarding the unique identifier, changing towards a more flexible client ID that is separate from the device ID. 
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+    - Additional algorithms MAY be supported, but the above list defines the minimum baseline for interoperability.
+    - The server MUST support all required algorithms (see above), so it can verify any compliant client, while clients MAY choose any of the required algorithms that fits their hardware capabilities.
+
+> Note: Further investigation and clarity to come regarding Margo's position on the root CA mentioned. 
```
This note seems out of place. There is no mention of the CA in the section right above this note.

ajcraig replied:
This has been removed in a latest commit.

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
-3. Generate a SHA-256 hash value for the requests's body
-4. Ensure the generated hash value matches the hash value from the message
+#### Security and Integrity Information
+Due to the limitations of utilizing mTLS with common OT infrastructure components, such as TLS terminating HTTPS load-balancer or a HTTPS proxy doing lawful inspection, Margo has adopted a certificate-based payload signing approach to protect payloads from being tampered with. By utilizing the certificates to create payload envelopes (HTTP Request body), the device's management client can ensure  ...
```
```suggestion
Due to the limitations of utilizing mTLS with common OT infrastructure components, such as TLS-terminating HTTPS load-balancer or a HTTPS proxy doing lawful inspection, Margo has adopted a certificate-based payload signing approach to protect payloads from being tampered with. By utilizing the certificates to create payload envelopes (HTTP Request body), the device's management client can ensure secure transport between the device's management client and the Workload Fleet Management's web service.
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+#### Security and Integrity Information
+Due to the limitations of utilizing mTLS with common OT infrastructure components, such as TLS terminating HTTPS load-balancer or a HTTPS proxy doing lawful inspection, Margo has adopted a certificate-based payload signing approach to protect payloads from being tampered with. By utilizing the certificates to create payload envelopes (HTTP Request body), the device's management client can ensure  ...
+
+- For API security, Server side TLS 1.3 (minimum) is used, where the keys are obtained from the Server's X.509 Certificate as defined in standard HTTP over TLS
```
```suggestion
- For API security, server-side TLS 1.3 (minimum) is used, where the keys are obtained from the Server's X.509 Certificate as defined in standard HTTP over TLS
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+Due to the limitations of utilizing mTLS with common OT infrastructure components, such as TLS terminating HTTPS load-balancer or a HTTPS proxy doing lawful inspection, Margo has adopted a certificate-based payload signing approach to protect payloads from being tampered with. By utilizing the certificates to create payload envelopes (HTTP Request body), the device's management client can ensure  ...
+
+- For API security, Server side TLS 1.3 (minimum) is used, where the keys are obtained from the Server's X.509 Certificate as defined in standard HTTP over TLS
+- For API integrity, the device's management client is issued a client specific X.509 certificate.
```
```suggestion
- For API integrity, the device's management client is issued a client-specific X.509 certificate.
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+- For API security, Server side TLS 1.3 (minimum) is used, where the keys are obtained from the Server's X.509 Certificate as defined in standard HTTP over TLS
+- For API integrity, the device's management client is issued a client specific X.509 certificate.
+- The issuer of the client X.509 certificate is trusted under the assumption that the root CA download to the Workload Fleet Management server occurs as a precondition to onboarding the devices 
+- Similarly the issuer of the server X.509 certificate is  trusted under the assumption that the root CA download to the device's management client occurs over a "protected" connection as part of the yet to be defined device onboarding procedure
```
```suggestion
- Similarly, the issuer of the server X.509 certificate is  trusted under the assumption that the root CA download to the device's management client occurs over a "protected" connection as part of the yet to be defined device onboarding procedure
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+#### Device Management Client
+Once the device management client has a message prepared for the Workload Fleet Management's web service, it MUST establish message integrity as defined in RFC 9421 by performing the following steps:
+
+- The device's management client MUST generate a SHA256 digest of the HTTP request body. Encode the digest in Base64 and include in the Content-Digest header. 
```
```suggestion
- The device's management client MUST generate a SHA256 digest of the HTTP request body. Encode the digest in Base64 and include it in the Content-Digest header. 
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
````diff
@@ -1,69 +1,137 @@
+        sig1=:<base64(signature)>:
+    ```
+
+> Note: The server MUST use the 'created' value to detect and prevent replay attacks. 
````
The meaning of this note isn't clear

matlec replied:
Agreed. I tried to address this by providing a little more context. Please review :)

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+> Note: The server MUST use the 'created' value to detect and prevent replay attacks. 
+
+#### Workload Fleet Manager Web-Service           
+- On receiving the message from the Device Client, The Workload Fleet Management's web service MUST do the following :
```
```suggestion
- On receiving the message from the Device Client, the Workload Fleet Management's web service MUST do the following :
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+#### Workload Fleet Manager Web-Service           
+- On receiving the message from the Device Client, The Workload Fleet Management's web service MUST do the following :
+    - It looks up the client certificate from the Client-ID in the API Request URL 
+    - The Workload Fleet Management's web service reads the following from the HTTP Request Header :
```
```suggestion
    - The Workload Fleet Management's web service reads the following from the HTTP Request Header:
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+        - Signature-Input
+        - Signature
+        - Content-Digest (if body is present)
+    - Use the Signature-Input in the header to determine which components were signed. Reconstruct the Signature Base canonical string using the actual values from the request including the SHA256 encoded content-digest from the received request body 
```
```suggestion
    - Use the Signature-Input in the header to determine which components were signed. Reconstruct the Signature Base canonical string using the actual values from the request, including the SHA256 encoded content-digest from the received request body 
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+        - Signature
+        - Content-Digest (if body is present)
+    - Use the Signature-Input in the header to determine which components were signed. Reconstruct the Signature Base canonical string using the actual values from the request including the SHA256 encoded content-digest from the received request body 
+    - Then extract the base64-encoded message signature from the Signature header, and also verifies the message signature string using the client's X.509 public-key.
```
```suggestion
    - Then extract the base64-encoded message signature from the Signature header and verifies the message signature string using the client's X.509 public-key.
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+        - Content-Digest (if body is present)
+    - Use the Signature-Input in the header to determine which components were signed. Reconstruct the Signature Base canonical string using the actual values from the request including the SHA256 encoded content-digest from the received request body 
+    - Then extract the base64-encoded message signature from the Signature header, and also verifies the message signature string using the client's X.509 public-key.
+    - If the message signature in the HTTP Header and the verified message signature match, then the payload is then processed by the Workload Fleet Management's web service.
```
```suggestion
    - If the message signature in the HTTP Header and the verified message signature match, then the payload is processed by the Workload Fleet Management's web service.
```

phil-abb commented on system-design/specification/margo-management-interface/api-requirements-and-security.md:
```diff
@@ -1,69 +1,137 @@
+    - Use the Signature-Input in the header to determine which components were signed. Reconstruct the Signature Base canonical string using the actual values from the request including the SHA256 encoded content-digest from the received request body 
+    - Then extract the base64-encoded message signature from the Signature header, and also verifies the message signature string using the client's X.509 public-key.
+    - If the message signature in the HTTP Header and the verified message signature match, then the payload is then processed by the Workload Fleet Management's web service.
+    - If the two do not match, the Workload Fleet Manager will respond with HTTP Error 401 as given below, and discontinue the session
```
Is 401 the correct HTTP Status code here? This is generally for unauthorized/unauthenticated responses, and I don't think this is related to authentication/authorization.

phil-abb commented on system-design/specification/margo-management-interface/certificate-api.md:
```diff
@@ -1,14 +1,23 @@
 # Certificate API
 
-In order to facilitate secure communication between the device's management client and workload orchestration web service the workload orchestration web service's root CA certificate must be downloaded using the Onboarding API's `certificate` endpoint.
+In order to facilitate secure communication between the device's management client and the Workload Fleet Manager, the device must first retrieve the root CA certificate  using the Onboarding API's `certificate` endpoint.
```
```suggestion
In order to facilitate secure communication between the device's management client and the Workload Fleet Manager, the device must first retrieve the root CA certificate using the Onboarding API's `certificate` endpoint.
```

phil-abb commented on system-design/specification/margo-management-interface/certificate-api.md:
```diff
@@ -1,14 +1,23 @@
 
-### Route and HTTP Methods
+- This endpoint MAY only be used to retrieve the root CA over an already trusted channel.
+- If no trusted channel exists, the root CA MUST be provided via out-of-band method. 
```
```suggestion
- If no trusted channel exists, the root CA MUST be provided via an out-of-band method. 
```

phil-abb commented on system-design/specification/margo-management-interface/deployment-status.md:
```diff
@@ -1,60 +1,75 @@
+While applying a new desired state, the device's management client MUST provide the Workload Fleet Manager service with an indication of the current workload deployment status. This is done by calling the Device API's `deployment status` endpoint.
 
-### Route and HTTP Methods
+- Requests to this endpoint MUST be authenticated using the HTTP Message Signature method as defined in the [Payload Security](../margo-management-interface/api-requirements-and-security.md#payload-security-method) section.
```
Is the really authentication?

phil-abb replied:
I'm going to resolve this comment. I don't think it's worth splitting hairs right now.

phil-abb commented on system-design/specification/margo-management-interface/deployment-status.md:
```diff
@@ -1,60 +1,75 @@
+| message   | string    | Y    | Associated error message that provides further details to the WFM about the error that was encountered. |
 
-### Example Request
+> Note: Error codes adn messages are implementation specific. 
```
```suggestion
> Note: Error codes and messages are implementation-specific. 
```

phil-abb replied:
The meaning of this note isn't clear

phil-abb commented on system-design/specification/margo-management-interface/deployment-status.md:
```diff
@@ -1,60 +1,75 @@
 
-### Route and HTTP Methods
+- Requests to this endpoint MUST be authenticated using the HTTP Message Signature method as defined in the [Payload Security](../margo-management-interface/api-requirements-and-security.md#payload-security-method) section.
+
```
One thing that is not addressed here is what should happen with a device that is not connected to the WFM when this is being done. We can probably just put a note for now, indicating this needs to be defined for devices that don't have a permanent connection.

nilanjan-samajdar replied:
Yes, the 'Poke' API from WFM to Device has not made it to the current draft. We might say that the API currently supports Devices which are functioning in connected mode to WFM.

ajcraig replied:
Note added here in the latest commit. Will address the poke strategy at a later point.

phil-abb commented on system-design/specification/margo-management-interface/device-capabilities.md:
```diff
@@ -1,28 +1,40 @@
 To ensure the WFM is kept up to date, the device's client MUST send updated capabilities information if any changes occur to the information originally provided (i.e., additional memory is added to the device).
 
-### Route and HTTP Methods
+- Requests to this endpoint MUST be authenticated using the HTTP Message Signature method as defined in the [Payload Security](../margo-management-interface/api-requirements-and-security.md#payload-security-method) section.
```
Maybe I'm being nitpicky, but to me, payload signing is not the same as authentication.

phil-abb replied:
I'm going to resolve this comment. I don't think it's worth splitting hairs right now.

phil-abb commented on system-design/specification/margo-management-interface/device-capabilities.md:
```diff
@@ -109,12 +121,12 @@ These enumerations are used as vocabularies for attribute values of the `DeviceC
 
 
-### Example Request
+## Example Device Capabilities Manifest Request
```
```suggestion
## Example Device Capabilities Payload
```

phil-abb commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
+    participant Client
+    participant Server
+
+    Note over Client,Server: 🔐 Initial Trust Establishment. 
```
```suggestion
    Note over Client,Server: 🔐 Initial Trust Establishment
```

phil-abb commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
+    Note over Client,Server: 🔐 Initial Trust Establishment. 
+    Note over Client,Server: The device's management client obtains the WFM's root CA certificate either out-of-band or via the Certificate API over an existing trusted channel.
+    Client->>Server: GET /onboarding/certificate
+    Server-->>Client: base64-encoded Root CA certificate
```
```suggestion
    Server-->>Client: Base64-encoded Root CA certificate
```

phil-abb commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
+    Note over Client,Server: The device's management client obtains the WFM's root CA certificate either out-of-band or via the Certificate API over an existing trusted channel.
+    Client->>Server: GET /onboarding/certificate
+    Server-->>Client: base64-encoded Root CA certificate
+    Client-->>Client: injecting Root CA into trusted store
```
```suggestion
    Client-->>Client: Injecting Root CA into trusted store
```

phil-abb commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
+    else Client rejected
+        Server-->>Client: 400 Bad Request {error: "Invalid certificate"}
+    else Client rejected
+        Server-->>Client: 403 Forbidden{ error: "Client rejected"}
```
```suggestion
        Server-->>Client: 403 Forbidden {error: "Client rejected"}
```

phil-abb commented on system-design/specification/margo-management-interface/device-onboarding.md:
```diff
@@ -1,47 +1,102 @@
+    WFMUser->>Server: Opportunity for user to approve or reject client onboarding
+    alt Public key trusted/user approved
+        Note over Server: Server verifies client certificate and assigns UUID
+        Server-->>Client: 201 Created { clientId }
```
```suggestion
        Server-->>Client: 201 Created {clientId}
```

phil-abb commented on system-design/specification/margo-management-interface/device-client-onboarding.md:
```diff
@@ -1,47 +1,102 @@
+| Fields       | Type            | Required?       | Description     |
+|-----------------|-----------------|-----------------|-----------------|
+| apiVersion      | string    | Y    | Identifier of the version the API resource follows.|
+| kind            | string    | Y    | Must be `OnboardingRequest`.|
```
What is the `kind` field for? If it's always "OnboardingRequest" why is it needed? Are there some future changes that will make use of this field? If so, it might be good to add a note about it.

matlec replied:
I agree. My general note was actually referring to this:

> Is there any rationale for how we choose to which apiVersion a specific Margo resource belongs? Today, it looks like each kind gets its own apiVersion which defeats the purpose of having apiVersion properties in the first place.

ajcraig replied:
This will be addressed post PR1. See latest comment in the PR. https://github.com/margo/specification/pull/124#issuecomment-3633181716

Commit:
Apply suggestions from code review

Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Co-authored-by: Philip Presson <philip.presson@us.abb.com>
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
- Changed names of files based on fb.
- Fixed various mkdocs build issues
- updated a couple pages for consistency within the concepts section
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

matlec committed:
fix: clarify link between Desired State and Deployment Status API

Signed-off-by: Matthias Lechner <matlec.public@gmail.com>

matlec committed:
fix: match style of desired state API docs with rest of the spec

Signed-off-by: Matthias Lechner <matlec.public@gmail.com>

matlec committed:
fix: clarify logging requirements for security events in manifest updates

Signed-off-by: Matthias Lechner <matlec.public@gmail.com>

matlec committed:
fix: clarify replay attack prevention details in signature handling

Signed-off-by: Matthias Lechner <matlec.public@gmail.com>

nilanjan-samajdar commented on system-design/concepts/workload-fleet-managers/device-capabilities.md:
The use of the word MUST has been removed in the explanation, which may dilute the need for a device to support this API

ajcraig commented:
> General:
> 
> 1. Is there any rationale for how we choose to which `apiVersion` a specific Margo resource belongs? Today, it looks like each kind gets its own apiVersion which defeats the purpose of having `apiVersion` properties in the first place.

We plan to update the apiVersion + the API grouping of the CRDs after the PR1 release. 
- Will remove the kind and apiversion from onboarding api / capabilities / deployment status
- Will rename the group to workloadmanagement.margo.org/v1alpha1 for both Deployment specification and Application description. 
- API specification will follow the Margo specification Symantec versioning. 
- API endpoints will stay v1 until we have a newer version with breaking changes.

Commit:
- Removed note on root CA that was out of place.
- Added note regarding disconnected/intermittent devices.
- changed version of open api specification to match paths version.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>

Commit:
Updated version of the API Swagger document along with the link.
Signed-off-by: Armand Craig <ajcraig@ra.rockwell.com>
